interfaces/cython/cantera/thermo.pyx

```diff
@@ -745,91 +747,250 @@ cdef class ThermoPhase(_SolutionBase):
+        self.thermo.setEquivalenceRatio_X(phi, &f[0], &o[0])
+
+    def set_mixture_fraction(self, mixFrac, fuel, oxidizer, basis='mole'):
+        """
+        Set the composition to a mixture of *fuel* and *oxidizer* at the
+        specified Bilger mixture fraction *mixFrac* (kg fuel / kg mixture), holding
+        temperature and pressure constant. Considers the oxidation of C to CO2,
+        H to H2O and S to SO2. Other elements are assumed not to participate in
+        oxidation (i.e. N ends up as N2). The *basis* detrmines the composition 
+        of fuel and oxidizer: basis='mole' (default) means mole fractions,
+        basis='mass' means mass fractions::
+
+            >>> gas.set_mixture_fraction(0.5, 'CH4', 'O2:1.0, N2:3.76')
+            >>> gas.mass_fraction_dict()
+            {'CH4': 0.5, 'N2': 0.38350014242997776, 'O2': 0.11649985757002226}
+            >>> gas.set_mixture_fraction(0.5, {'NH3':0.8, 'CO':0.2}, 'O2:1.0')
+            >>> gas.mass_fraction_dict()
+            {'CO': 0.14568206877899598, 'NH3': 0.354317931221004, 'O2': 0.5}
+
+        :param mixFrac: Mixture fraction (kg fuel / kg mixture)
+        :param fuel:
+            Fuel species name or mole/mass fractions as string, array, or dict.
+        :param oxidizer:
+            Oxidizer species name or mole/mass fractions as a string, array, or
+            dict.
+        :param basis: determines if *fuel* and *oxidizer* are given in mole
+                      fractions (basis='mole') or mass fractions (basis='mass')
+      
+        """
 
-        nO = np.array([self.n_atoms(k, 'O') for k in range(self.n_species)])
+        if (isinstance(fuel, str) and ':' not in fuel
+            and fuel in self.species_names):
+            fuel += ':1.0'
 
-        if 'C' in self.element_names:
-            nC = np.array([self.n_atoms(k, 'C') for k in range(self.n_species)])
-        else:
-            nC = np.zeros(self.n_species)
+        if (isinstance(oxidizer, str) and ':' not in oxidizer
+            and oxidizer in self.species_names):
+            oxidizer += ':1.0'
 
-        if 'H' in self.element_names:
-            nH = np.array([self.n_atoms(k, 'H') for k in range(self.n_species)])
-        else:
-            nH = np.zeros(self.n_species)
+        original_state = self.state
+        
+        if basis == 'mole':
+            self.TPX = None, None, fuel
+            Xf = self.X
+            self.TPX = None, None, oxidizer
+            Xo = self.X
+        elif basis == 'mass':
+            self.TPY = None, None, fuel
+            Xf = self.X
+            self.TPY = None, None, oxidizer
+            Xo = self.X
+
+        cdef np.ndarray[np.double_t, ndim=1] f = \
+                np.ascontiguousarray(Xf, dtype=np.double)
+        cdef np.ndarray[np.double_t, ndim=1] o = \
+                np.ascontiguousarray(Xo, dtype=np.double)
+        
+        self.state = original_state 
+        
+        self.thermo.setMixtureFraction_X(mixFrac, &f[0], &o[0])
```

Comment:
Watch out for end-of-line whitespace, here and in several other places. I recommend configuring your editor to highlight it so it can be avoided.
```suggestion

        self.state = original_state

        self.thermo.setMixtureFraction_X(mixFrac, &f[0], &o[0])
```